File: observer/observer.py
# Define the Subject (Observable) interface
class WeatherStation:
    def add_observer(self, observer):
        pass

    def remove_observer(self, observer):
        pass

    def notify_observers(self):
        pass

# Concrete implementation of Subject
class ConcreteWeatherStation(WeatherStation):
    def __init__(self):
        self.temperature = 0
        self.observers = []

    def add_observer(self, observer):
        self.observers.append(observer)

    def remove_observer(self, observer):
        self.observers.remove(observer)

    def notify_observers(self):
        for observer in self.observers:
            observer.update(self.temperature)

    def set_temperature(self, temperature):
        self.temperature = temperature
        self.notify_observers()

# Define the Observer interface
class WeatherDisplay:
    def update(self, temperature):
        pass

# Concrete implementation of Observer
class CurrentConditionsDisplay(WeatherDisplay):
    def update(self, temperature):
        print(f"Current conditions: {temperature}°C")

# Client code
weather_station = ConcreteWeatherStation()
current_display = CurrentConditionsDisplay()

# Registering the observer
weather_station.add_observer(current_display)

# Changing the temperature (this will trigger observer notifications)
weather_station.set_temperature(25)
weather_station.set_temperature(30)

File: observer/observer.kt
// Subject
interface WeatherStation {
    fun addObserver(observer: WeatherDisplay)
    fun removeObserver(observer: WeatherDisplay)
    fun notifyObservers()
}

// Concrete Subject
class ConcreteWeatherStation : WeatherStation {
    private var temperature: Float = 0.0f
    private val observers = mutableListOf<WeatherDisplay>()

    override fun addObserver(observer: WeatherDisplay) {
        observers.add(observer)
    }

    override fun removeObserver(observer: WeatherDisplay) {
        observers.remove(observer)
    }

    override fun notifyObservers() {
        for (observer in observers) {
            observer.update(temperature)
        }
    }

    fun setTemperature(newTemperature: Float) {
        this.temperature = newTemperature
        notifyObservers()
    }
}

// Observer
interface WeatherDisplay {
    fun update(temperature: Float)
}

// Concrete Observer
class CurrentConditionsDisplay : WeatherDisplay {
    override fun update(temperature: Float) {
        println("Current conditions: $temperature °C")
    }
}

// Client Code
fun main() {
    val weatherStation = ConcreteWeatherStation()
    val currentConditions = CurrentConditionsDisplay()

    weatherStation.addObserver(currentConditions)

    println("Changing temperature...")
    weatherStation.setTemperature(30.0f)

    println("Changing temperature again...")
    weatherStation.setTemperature(32.0f)
}
